src/UpdateCapableInterface.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Dhii\Storage\Resource;
+
+use Dhii\Expression\LogicalExpressionInterface;
+use Psr\Container\ContainerInterface;
+use Traversable;
+
+/**
+ * Something that can update records in storage.
+ *
+ * @since [*next-version*]
+ */
+interface UpdateCapableInterface
+{
+    /**
+     * Updates the records in storage.
+     *
+     * @since [*next-version*]
+     *
+     * @param array                           $changeSet An array of changes, mapping field name to changed value.
```

Comment:
Is there a reason why this is not `Traversable`?

Reply:
Can't remember the reason for changing this to just `array`.

I don't `Traversable` should, or can, be supported. Implementations need the mapping from field to new value. If anything, this is expected to be _at least_ a container, i.e. `array|ArrayAccess|stdClass|Container`. Thoughts?

Reply:
Show me how the mapping is done. Iterating over known mapping configuration, and retrieving keys from the set? Or iterating over the set, and retrieving mappings?

Reply:
In my current abstract implementation, a list of known field names is iterated over. This list is something that the abstract RM is aware of, or at least has access to. Each field is used as a key to read from this changeset, thus ensuring that only correct fields are used.

In code, this translates to

```
foreach ($fields as $_field) {
    $_changedValue = $changeSet[$_field];
	// ...
}
```

Reply:
Of course, this is an implementation detail. Perhaps the interface shouldn't assume _how_ it will be read. Maybe just depend on `MapInterface`?

Reply:
Why not the other way around? And what happens with the WP post model, which will treat all unknown keys as meta keys?

Reply:
Yeah ..... I think this is why I changed these to arrays, so that WP RMs can be extracted nicely such that known post data can be extracted and then be intersected to get the remainder.

Reply:
Looking at the code for WP native insertion RM, this is exactly what's happening. It wouldn't be as clean if the list is iterated and both post data AND post meta are extracted in one iteration, but it would certainly make everything more interoperable through container helpers.

Reply:
That doesn't answer the question "why?". Is there a reason why not traverse the set, and get the mappings from the config?

Reply:
What is exactly what's happening?

Reply:
Actually now that I think about it, this approach was most likely selected due to performance reasons. Iterating over the changeset and each time checking if a key is in a list of known post keys is expensive.

Reply:
Probably not more expensive than checking every time if the key is in the set. Unless the config mapping is the other way around, i.e. columns map to keys.

Reply:
No, it's exactly as it is for the PDO/WPDB RMS. The list of known post keys is iterated, and if the changeset has the key, the value is taken.

```
$postData = [];
foreach ($knownPostKeys as $_key) {
	if (!isset($changeset[$_key])) {
		continue;
	}
	$postData[$_key] = $changeset[$_key];
}
```

Where `$knownPostKeys` comes from an abstract method and might look something like this: `["post_ID", "post_title", "post_author", ...]`

The above is DEFINITELY not more expensive than

```
foreach ($changeset as $_key => $_value) {
    if (in_array($knownPostKeys, $_key) { // <=
		//  then post data
	} else {
		// then post meta
	}
}
```

Reply:
Don't misunderstand please. I'm not defending the above implementation.

I'd rather not open up the param type and somehow find a way to improve this implementation, than to restrict the API just because of this implementation's difficulty to achieve the desired result.

Reply:
As far as I understood from this, you agree that some re-factoring is in order.

Reply:
Feel the Dhii in you, Jedi...

Reply:
I've been feeling the re-factoring coming when I pushed my local changes, only to discover that I had unpushed committed changes. I was like "waaaaaat is this?". And I didn't like what I saw. But now I can see why I did it. Everything always comes down to WP. sigh...

What do you think about the implementation though? Take a look at the one that I remarked as being more expensive. Do you see why it is expensive?

What do you think about changing it such that the `$knownPostKeys` is a map and not a list?

This would also mean that the WP update RM can, instead of expecting the changeset to be keyed by WP post data keys (such as `post_ID`, `post_title`, etc...), expect it to be keyed by "friendly" field names (like say `id`, `title`, etc...).

Thoughts?

Reply:
No, don't really see why it's more expensive. The latter looks like it can do both in one go.

Yes, can use friendly field names, sounds good to me. However, please, if you find that there's a lot to change, let me know.